src/PlayerAI.cpp

```diff
@@ -141,7 +141,25 @@ TapNoteScore PlayerAI::GetTapNoteScore( const PlayerState* pPlayerState )
 
 	TapScoreDistribution& distribution = g_Distributions[iCpuSkill];
 
-	return distribution.GetTapNoteScore();
+	TapNoteScore score = distribution.GetTapNoteScore();
+
+	const auto& disabledWindows = pPlayerState->m_PlayerOptions.GetCurrent().m_twDisabledWindows;
+	for (int i = score; i >= TNS_W5; i--)
+	{
+		TapNoteScore cur_score = (TapNoteScore)i;
+		// Downgrade the TapNoteScore if that specific window is disabled.
+		if (cur_score == TNS_W1 && disabledWindows[TW_W1])
+			score = TNS_W2;
+		if (cur_score == TNS_W2 && disabledWindows[TW_W2])
+			score = TNS_W3;
+		if (cur_score == TNS_W3 && disabledWindows[TW_W3])
+			score = TNS_W4;
+		if (cur_score == TNS_W4 && disabledWindows[TW_W4])
+			score = TNS_W5;
+		if (cur_score == TNS_W5 && disabledWindows[TW_W5])
+			score = TNS_None;
+	}
```

Comment:
I think the loop is not required. This cascade without the loop should already do the trick:
```c++
if (score == TNS_W1 && disabledWindows[TW_W1])
	score = TNS_W2;
if (score == TNS_W2 && disabledWindows[TW_W2])
	score = TNS_W3;
if (score == TNS_W3 && disabledWindows[TW_W3])
	score = TNS_W4;
if (score == TNS_W4 && disabledWindows[TW_W4])
	score = TNS_W5;
if (score == TNS_W5 && disabledWindows[TW_W5])
	score = TNS_None;
```

Reply:
Done